Fixed links to ERC-721, added keyword "payable" in some functions, and update to review (#5155)
ethereum/EIPs: EIPS/eip-4519.md

eip: 4519
title: Non-Fungible Tokens Tied to Physical Assets (SmartNFT)

> description: Standard interface for non-fungible tokens representing physical assets that can generate or recover their own accounts and obey users.

Reviewer: We now try to discourage "standard" in titles and description (since most EIPs are standards for something or other):

```suggestion
description: Interface for non-fungible tokens representing physical assets that can generate or recover their own accounts and obey users.
```

description: Standard interface for non-fungible tokens representing physical assets that can generate or recover their own accounts and obey users.

> author: Javier Arcenegui (@Hardblock-IMSE-CNM), Rosario Arjona <arjona@imse-cnm.csic.es>, Roberto Román <roman@imse-cnm.csic.es>, Iluminada Baturone <lumi@imse-cnm.csic.es>

Reviewer 2: Is it possible to get more authors' GitHub handles in this? They are preferred to emails.

> author: Javier Arcenegui (@Hardblock-IMSE-CNM), Rosario Arjona <arjona@imse-cnm.csic.es>, Roberto Román <roman@imse-cnm.csic.es>, Iluminada Baturone <lumi@imse-cnm.csic.es>

Author: ```suggestion
author: Javier Arcenegui (@Hardblock-IMSE-CNM), Rosario Arjona (@RosarioArjona), Roberto Román <roman@imse-cnm.csic.es>, Iluminada Baturone (@lumi2018)
```

discussions-to: https://ethereum-magicians.org/t/new-proposal-of-smart-non-fungible-token/7677

> status: Review

Reviewer 2: I can't make comments on parts that are unchanged, so I'll be putting them here:

- L32-L40: The images do not appear to be correctly loaded, would you mind double-checking the links?
- L56: I believe you have an extra space in `engagedWithOwner`
- Throughout: You are using quotation marks to refer to code/states. It's recommended to use backticks (\`) instead. Here's an example:
  
  L203: ...some functions such as "startUserEngagement" are only...
  L203 (with backticks): ...some functions such as `startUserEngagement` are only...
- Throughout: There are quite a few grammar errors. I would recommend running the paper through a grammar checker and making any appropriate changes.

**SmartNFT**
This EIP proposes a Non-Fungible Token tied to a physical asset. The asset is able to generate an Ethereum account and authenticate its user and its owner. Hence, the asset can be considered as a smart asset associated with an NFT. If the asset and the token are regarded as one thing, we can talk about a smart NFT. The demand for smart NFTs with user management or with a tie to a physical asset are growing (for example, in the context of the Internet of Things). Therefore, it is essential to establish a standard capable of including all these options working together or separately. The incorporation of an Ethereum address of the user or an Ethereum address of a physical asset to the smart NFT is optional. However, it does not make sense that the smart NFT does not include any of the two Ethereum addresses because, in that case, the NFT would be an ERC-721 token. The possibility of generating two separate interfaces has been proposed, however, some functions such as `startUserEngagement` are only be available if both options are implemented, so a single interface with all the options is proposed.

**Authentication**

> This EIP proposes using the Smart Contract to verify the mutual authentication process between the physical asset and the owner or the user by verifying the hash of a shared key.

Reviewer 2: ```suggestion
This EIP proposes using the smart contract to verify the mutual authentication process between the physical asset and the owner or the user by verifying the hash of a shared key.
```

Smart contract isn't a proper noun.

    function userBalanceOfAnOwner(address _addressUser, address _addressOwner) external view returns (uint256);
}
```

> This interface is an extension of the [ERC-721](./eip-721.md), is compatible with the standard, and needs the ERC-721 interface. Then, metadata and enumeration extensions are compatible and included in this draft.
> Like EIP-165 is required, supportsInterface SHOULD BE overridden to include the interfaceId of this EIP.

Reviewer 2: ```suggestion
```

This should be evident from the provided solidity interface.

> This EIP proposes the attributes timestamp (to register in the blockchain whenever the physical asset checks the tie with its token) and timeout (to register the maximum delay time established for the physical asset to prove again the tie). These attributes avoid a malicious owner or user could use the asset endlessly.

Reviewer 2: Is the timestamp measured in blocks, seconds, or milliseconds?

Reviewer 2: Would you mind specifying that?

Also, I would much prefer if `block.number` is used.

Author: We prefer to use `block.timestamp` since it is more difficult for an electronic device to estimate the number of blocks than the time elapsed since the last time the timestamp was updated.

**ERC-721-based**

> The [ERC-721](./eip-721.md) is a strong foundation for any NTF, nevertheless Ethereum addresses of assets and users, and authentication and key exchange processes are not considered. This EIP proposes an update of the [ERC-721](./eip-721.md) respecting backward compatibility.

Reviewer 2: ```suggestion
[EIP-721](./eip-721.md) is the most commonly-used standard for generic NFTs. This EIP extends ERC-721 for backward compatibility.
```

> ## Backwards Compatibility
> This implementation is an extension of the ERC-721 standard, then it is not only compatible with the standard, but also is a way to improve some of the current NFT tokens based on the ERC-721 standard.

Reviewer 2: ```suggestion
## Backwards Compatibility
This standard is an extension of ERC-721. It is fully compatible with both commonly-used optional extensions (`IERC721Metadata` and `IERC721Enumerable`) mentioned in the EIP-721 standard.

```

## Test Cases

> The test case presented in the paper shown below is addressed in **0x7eB5A03E7ED70ABf70fee48965D0411d37F335aC** in kovan testnet and the code is available [here](../assets/eip-4519/PoC_SmartNFT/README.md)

Reviewer 2: ```suggestion
The test cases presented in the paper shown below are available the assets folder [here](../assets/eip-4519/PoC_SmartNFT/README.md).
```

I would recommend removing the reference to the Kovan testnet address. Users should probably deploy their own test suites.

> This EIP proposes the attributes timestamp (to register in the blockchain whenever the physical asset checks the tie with its token) and timeout (to register the maximum delay time established for the physical asset to prove again the tie). These attributes avoid a malicious owner or user could use the asset endlessly.

Author: ```suggestion
This EIP proposes the attributes timestamp (to register in the blockchain whenever the physical asset checks the tie with its token) and timeout (to register the maximum delay time established for the physical asset to prove again the tie). These attributes avoid a malicious owner or user could use the asset endlessly.
When the asset calls `updateTimestamp`, the smart contract must call `block.timestamp`, that provide current block timestamp as seconds since unix epoch. For this reason, `timeout`  must be provided in seconds.
```

**ERC-721-based**
[EIP-721](./eip-721.md) is the most commonly-used standard for generic NFTs. This EIP extends ERC-721 for backwards compatibility.
  
## Backwards Compatibility
This standard is an extension of ERC-721. It is fully compatible with both of the commonly used optional extensions (`IERC721Metadata` and `IERC721Enumerable`) mentioned in the EIP-721 standard.

## Test Cases

> The test cases presented in the paper shown below are available the assets folder [here](../assets/eip-4519/PoC_SmartNFT/README.md).

Reviewer 2: ```suggestion
The test cases presented in the paper shown below are available the assets folder [here](../assets/eip-4519/PoC_SmartNFT/README.md).

```

Add whitespace
